File: Data Structures & Algorithms/design-circular-queue/submission-0.java
class Node{
    int val;
    Node next, prev;

    Node(int val, Node next, Node prev){
        this.val = val;
        this.next = next;
        this.prev = prev;
    }
}

class MyCircularQueue {
    private int size;
    private int cap;
    private Node left, right;

    public MyCircularQueue(int k) {
        size = 0;
        cap = k;
        left = new Node(0, null, null);
        right = new Node(0, null, null);

        left.next = right;
        right.prev = left;
    }

    public boolean enQueue(int value) {
        if(isFull()){
            return false;
        }

        Node curr = new Node(value, null, null);
        Node rPrev = right.prev;
        rPrev.next = curr;
        curr.next = right;
        curr.prev = rPrev;

        size++;

        return true;
    }
    
    public boolean deQueue() {
        if(isEmpty()){
            return false;
        }

        Node currLast = left.next.next;
        currLast.prev = left;
        left.next = currLast;

        size--;

        return true;
        
    }
    
    public int Front() {
        return right.prev.val;
    }
    
    public int Rear() {
        return left.next.val;
    }
    
    public boolean isEmpty() {
        return size == 0;
    }
    
    public boolean isFull() {
        return size == cap;
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue obj = new MyCircularQueue(k);
 * boolean param_1 = obj.enQueue(value);
 * boolean param_2 = obj.deQueue();
 * int param_3 = obj.Front();
 * int param_4 = obj.Rear();
 * boolean param_5 = obj.isEmpty();
 * boolean param_6 = obj.isFull();
 */

File: Data Structures & Algorithms/populating-next-right-pointers-in-each-node/submission-0.js
/**
 * Definition for a binary tree node.
 * class Node {
 *     constructor(val = 0, left = null, right = null, next = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {Node} root
     * @return {Node}
     */

    // []
    // [1]
    // 1 -> [2, 3], 

    connect(root) {
        if(root == null){
            return root;
        }

        const q = new Queue();
        let prevNode = null;

        q.push(root);

        while(q.size() > 0){
            const len = q.size();

            for(let i = 0; i < len; i++){
                const curr = q.dequeue();

                if(curr.left){
                    q.enqueue(curr.left);
                    q.enqueue(curr.right);
                }

                if(prevNode){
                    prevNode.next = curr;
                }

                prevNode = curr;

                if(i === (len - 1)){
                    curr.next = null;
                    prevNode = null;
                }
            }
        }


        return root;
        
        
    }
}
